src/java.base/share/classes/javax/crypto/EncryptedPrivateKeyInfo.java

```
@@ -492,68 +501,110 @@ private static EncryptedPrivateKeyInfo encryptKeyImpl(PrivateKey key,
-     * @throws NullPointerException if {@code decryptKey} is null
+     * @return a {@code PrivateKey} extracted from the encrypted data
+     * @throws GeneralSecurityException if an error occurs parsing,
+     * decrypting the data, or producing the key object.
+     * @throws NullPointerException if {@code decryptKey} is {@code null}
      *
      * @since 25
      */
     @PreviewFeature(feature = PreviewFeature.Feature.PEM_API)
     public PrivateKey getKey(Key decryptKey, Provider provider)
         throws GeneralSecurityException {
         Objects.requireNonNull(decryptKey,"decryptKey cannot be null.");
-        PKCS8EncodedKeySpec p = getKeySpecImpl(decryptKey, provider);
-        try {
-            if (provider == null) {
-                return KeyFactory.getInstance(
-                    KeyUtil.getAlgorithm(p.getEncoded())).
-                    generatePrivate(p);
-            }
-            return KeyFactory.getInstance(KeyUtil.getAlgorithm(p.getEncoded()),
-                provider).generatePrivate(p);
-        } catch (IOException e) {
-            throw new GeneralSecurityException(e);
-        }
+        return PKCS8Key.parseKey(decryptData(decryptKey, provider), null);
+    }
+
+    /**
+     * Extracts and returns a {@code KeyPair} object from the encrypted data.
+     * The data must contain both a public and private key.
+     *
+     * @param password the password used in the PBE encryption.  This array
+     *                 will be cloned before being used.
+     * @return a {@code KeyPair} extracted from the encrypted data
+     * @throws GeneralSecurityException if an error occurs while parsing,
+     * decrypting the data, or producing the {@code KeyPair} object.
+     * @throws NullPointerException if {@code password} is {@code null}
+     *
+     * @since 26
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.PEM_API)
+    public KeyPair getKeyPair(char[] password) throws GeneralSecurityException {
```

This should specifically throw an `InvalidKeyException` if decryption fails. I think you want to specify the exact subclasses when it is clearly the right behavior. In this case, the method should always throw `InvalidKeyException` on decryption errors.

Also why generalize it to throw `GeneralSecurityException`? This should throw the same exceptions as `getKeySpec(Key, Provider)` since the params are the same.